README.md
# debug-es

[![Build Status](https://travis-ci.org/LightouchDev/debug-es.svg?branch=master)](https://travis-ci.org/LightouchDev/debug-es)
[![Coverage Status](https://coveralls.io/repos/github/LightouchDev/debug-es/badge.svg?branch=master)](https://coveralls.io/github/LightouchDev/debug-es?branch=master)
[![JavaScript Style Guide](https://cdn.rawgit.com/standard/standard/master/badge.svg)](https://github.com/standard/standard)

<img width="647" src="https://user-images.githubusercontent.com/71256/29091486-fa38524c-7c37-11e7-895f-e7ec8e1039b6.png">

A tiny JavaScript debugging utility modelled after Node.js core's debugging
technique. Works in Node.js and web browsers.

NOTE: it's a fork of [`debug`](https://github.com/visionmedia/debug) for es module development.

## Installation

```bash
$ npm install debug-es
or
$ yarn add debug-es
```

## Requirement

- ES6 support
  - node >= 6
  - node 4 (not tested)

## Usage

`debug-es` exposes a function; simply pass this function the name of your module, and it will return a decorated version of `console.error` for you to pass debug statements to. This will allow you to toggle the debug output for different parts of your module as well as the module as a whole.

```js
const debug = require('debug-es')('http')
const http = require('http')
const name = 'My App'

// fake app

debug('booting %o', name)

http.createServer((req, res) => {
  debug(`${req.method} ${req.url}`)
  res.end('hello\n')
}).listen(3000, () => debug('listening'))

// fake worker of some kind

require('./worker')
```

```js
const a = require('debug-es')('worker:a')
const b = require('debug-es')('worker:b')

function work() {
  a('doing lots of uninteresting work')
  setTimeout(work, Math.random() * 1000)
}

work()

function workb() {
  b('doing some work')
  setTimeout(workb, Math.random() * 2000)
}

workb()
```

The `DEBUG` environment variable is then used to enable these based on space or
comma-delimited names.

Here are some examples:

<img width="647" alt="screen shot 2017-08-08 at 12 53 04 pm" src="https://user-images.githubusercontent.com/71256/29091703-a6302cdc-7c38-11e7-8304-7c0b3bc600cd.png">
<img width="647" alt="screen shot 2017-08-08 at 12 53 38 pm" src="https://user-images.githubusercontent.com/71256/29091700-a62a6888-7c38-11e7-800b-db911291ca2b.png">
<img width="647" alt="screen shot 2017-08-08 at 12 53 25 pm" src="https://user-images.githubusercontent.com/71256/29091701-a62ea114-7c38-11e7-826a-2692bedca740.png">

### Choose implement (for tree-shaking)

by default, **debug-es** would do lots of runtime environment check to choose implement (**node** or **browser**). but it's impossible for tree-shaking without modifying code (or DefinePlugin in webpack)

now, we expose init function for tree-shaking.

#### CommonJS

```javascript
// for browser
const debug = require('debug-es/lib/browser')()

// for node
const debug = require('debug-es/lib/node')()

// both (no tree-shaking)
const debug = require('debug-es')
```

#### ES Module

ES modules are already **bundled with (non-es) deps**, it can be transpiled by babel.

```javascript
// for browser
import getDebug from 'debug-es/esm/browser'
const debug = getDebug()

// for node
import getDebug from 'debug-es/esm/node'
const debug = getDebug()

// both (no tree-shaking)
import debug from 'debug-es/esm'
```

### Windows command prompt notes

#### CMD

On Windows the environment variable is set using the `set` command.

```cmd
set DEBUG=*,-not_this
```

Example:

```cmd
set DEBUG=* & node app.js
```

#### PowerShell (VS Code default)

PowerShell uses different syntax to set environment variables.

```cmd
$env:DEBUG = "*,-not_this"
```

Example:

```cmd
$env:DEBUG='app';node app.js
```

Then, run the program to be debugged as usual.

npm script example:

```javascript
  "windowsDebug": "@powershell -Command $env:DEBUG='*';node app.js",
```

## Namespace Colors

Every debug instance has a color generated for it based on its namespace name.
This helps when visually parsing the debug output to identify which debug instance
a debug line belongs to.

### Web Browser

Colors are also enabled on "Web Inspectors" that understand the `%c` formatting
option. These are WebKit web inspectors, Firefox ([since version
31](https://hacks.mozilla.org/2014/05/editable-box-model-multiple-selection-sublime-text-keys-much-more-firefox-developer-tools-episode-31/))
and the Firebug plugin for Firefox (any version).

<img width="524" src="https://user-images.githubusercontent.com/71256/29092033-b65f9f2e-7c39-11e7-8e32-f6f0d8e865c1.png">

## Millisecond diff

When actively developing an application it can be useful to see when the time spent between one `debug()` call and the next. Suppose for example you invoke `debug()` before requesting a resource, and after as well, the "+NNNms" will show you how much time was spent between calls.

<img width="647" src="https://user-images.githubusercontent.com/71256/29091486-fa38524c-7c37-11e7-895f-e7ec8e1039b6.png">

When stdout is not a TTY, `Date#toISOString()` is used, making it more useful for logging the debug information as shown below:

<img width="647" src="https://user-images.githubusercontent.com/71256/29091956-6bd78372-7c39-11e7-8c55-c948396d6edd.png">

## Conventions

If you're using this in one or more of your libraries, you _should_ use the name of your library so that developers may toggle debugging as desired without guessing names. If you have more than one debuggers you _should_ prefix them with your library name and use ":" to separate features. For example "bodyParser" from Connect would then be "connect:bodyParser".  If you append a "*" to the end of your name, it will always be enabled regardless of the setting of the DEBUG environment variable.  You can then use it for normal output as well as debug output.

## Wildcards

The `*` character may be used as a wildcard. Suppose for example your library has
debuggers named "connect:bodyParser", "connect:compress", "connect:session",
instead of listing all three with
`DEBUG=connect:bodyParser,connect:compress,connect:session`, you may simply do
`DEBUG=connect:*`, or to run everything using this module simply use `DEBUG=*`.

You can also exclude specific debuggers by prefixing them with a "-" character.
For example, `DEBUG=*,-connect:*` would include all debuggers except those
starting with "connect:".

## Environment Variables

When running through Node.js, you can set a few environment variables that will
change the behavior of the debug logging:

| Name      | Purpose                                         |
|-----------|-------------------------------------------------|
| `DEBUG`   | Enables/disables specific debugging namespaces. |
| `DEBUG_HIDE_DATE` | Hide date from debug output (non-TTY).  |
| `DEBUG_COLORS`| Whether or not to use colors in the debug output. |
| `DEBUG_DEPTH` | Object inspection depth.                    |
| `DEBUG_SHOW_HIDDEN` | Shows hidden properties on inspected objects. |

__Note:__ The environment variables beginning with `DEBUG_` end up being
converted into an Options object that gets used with `%o`/`%O` formatters.
See the Node.js documentation for
[`util.inspect()`](https://nodejs.org/api/util.html#util_util_inspect_object_options)
for the complete list.

## Formatters

Debug uses [printf-style](https://wikipedia.org/wiki/Printf_format_string) formatting.
Below are the officially supported formatters:

| Formatter | Representation |
|-----------|----------------|
| `%O`      | Pretty-print an Object on multiple lines. |
| `%o`      | Pretty-print an Object all on a single line. |
| `%s`      | String. |
| `%d`      | Number (both integer and float). |
| `%j`      | JSON. Replaced with the string '[Circular]' if the argument contains circular references. |
| `%%`      | Single percent sign ('%'). This does not consume an argument. |

### Custom formatters

You can add custom formatters by extending the `debug.formatters` object.
For example, if you wanted to add support for rendering a Buffer as hex with
`%h`, you could do something like:

```js
const createDebug = require('debug-es')
createDebug.formatters.h = (v) => {
  return v.toString('hex')
}

// ...elsewhere
const debug = createDebug('foo')
debug('this is hex: %h', new Buffer('hello world'))
//   foo this is hex: 68656c6c6f20776f726c6421 +0ms
```

## Browser Support

Debug's enable state is currently persisted by `localStorage`.
Consider the situation shown below where you have `worker:a` and `worker:b`,
and wish to debug both. You can enable this using `localStorage.debug`:

```js
localStorage.debug = 'worker:*'
```

And then refresh the page.

```js
const a = debug('worker:a')
const b = debug('worker:b')

setInterval(function(){
  a('doing some work')
}, 1000)

setInterval(function(){
  b('doing some work')
}, 1200)
```

## Output streams

  By default `debug` will log to stderr, however this can be configured per-namespace by overriding the `log` method:

```js
const debug = require('debug-es')
const error = debug('app:error')

// by default stderr is used
error('goes to stderr!')

const log = debug('app:log')
// set this namespace to log via console.log
log.log = console.log.bind(console) // don't forget to bind to console!
log('goes to stdout')
error('still goes to stderr!')

// set all output to go via console.info
// overrides all per-namespace log settings
debug.log = console.info.bind(console)
error('now goes to stdout via console.info')
log('still goes to stdout, but via console.info now')
```

## Set dynamically

You can also enable debug dynamically by calling the `enable()` method :

```js
let debug = require('debug-es')

console.log(1, debug.enabler('test'))

debug.enable('test')
console.log(2, debug.enabler('test'))

debug.disable()
console.log(3, debug.enabler('test'))

```

print :

```plain
1 false
2 true
3 false
```

Usage :
`enable(namespaces)`
`namespaces` can include modes separated by a colon and wildcards.

Note that calling `enable()` completely overrides previously set DEBUG variable :

```bash
$ DEBUG=foo node -e 'const dbg = require("debug-es"); dbg.enable("bar"); console.log(dbg.enabler("foo"))'
=> false
```

## Checking whether a debug target is enabled

After you've created a debug instance, you can determine whether or not it is
enabled by checking the `enabled` property:

```javascript
const debug = require('debug-es')('http')

if (debug.enabled) {
  // do stuff...
}
```

You can also manually toggle this property to force the debug instance to be
enabled or disabled.

## Authors

- TJ Holowaychuk
- Nathan Rajlich
- Andrew Rhyne
- Miau Lightouch

## License

(The MIT License)

Copyright (c) 2014-2017 TJ Holowaychuk &lt;tj@vision-media.ca&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Changelog

## 1.2.3 / 2018-08-15

* Fix: use rollup to distribute module, make es module would work as expected with babel.
* Fix: init function without namespace would return new instance now.

## 1.2.2 / 2018-08-12

* Fix: prevent ci remove the files for deploy

## 1.2.1 / 2018-08-12

* Fix: forgot to build module before deploy

## 1.2.0 / 2018-08-12

* Revert: debug timebase
* Breaking: Re-write into ES module
* Addition: TVMLKit support
* Addition: Tree-shaking compatibility

## 1.1.0 / 2018-06-20

* Feature: configurable debug instance timebase
* Update: devDependencies
* Fix: new color test condition: unexpected useragent

## 1.0.0 / 2018-05-15

* Breaking: Re-write in ES6
* Breaking: createDebug.humanize move to env.humanize
* Breaking: createDebug.enabled renamed to createDebug.enabler
* Addition: nwjs support
* Addition: `process` package support
* Addition: 99% coverage test
* Fix: JSON.stringify circular issue
* Fix: Edge > 16.16215 support console colors
* Fix: window.webkitURL is existed in Edge issue


src/node.js
import tty from 'tty'
import util from 'util'
import supportsColor from 'supports-color'

import init from './init'

const inject = (createDebug) => ({
  /**
   * Colors.
   */
  colors: (() => {
    try {
      if (supportsColor.stderr && supportsColor.stderr.has256) {
        return [
          20, 21, 26, 27, 32, 33, 38, 39, 40, 41, 42, 43, 44, 45, 56, 57, 62,
          63, 68, 69, 74, 75, 76, 77, 78, 79, 80, 81, 92, 93, 98, 99, 112,
          113, 128, 129, 134, 135, 148, 149, 160, 161, 162, 163, 164, 165,
          166, 167, 168, 169, 170, 171, 172, 173, 178, 179, 184, 185, 196,
          197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209,
          214, 215, 220, 221
        ]
      }
    } catch (error) {}

    return [ 6, 2, 3, 4, 5, 1 ]
  })(),

  /**
   * Adds ANSI color escape codes if enabled.
   *
   * NOTE: only 'this' in formatArgs is scoped in 'debug' instance,
   *       not in 'createDebug' scope.
   *
   * @api public
   */
  formatArgs (args) {
    const name = this.namespace

    if (this.useColors) {
      const c = this.color
      const colorCode = '\u001b[3' + (c < 8 ? c : `8;5;${c}`)
      const prefix = `  ${colorCode};1m${name} \u001b[0m`

      args[0] = prefix + args[0].split('\n').join(`\n prefix`)
      args.push(`${colorCode}m+${createDebug.humanize(this.diff)}\u001b[0m`)
    } else {
      const date = createDebug.inspectOpts.hideDate
        ? ''
        : new Date().toISOString() + ' '
      args[0] = `${date}${name} ${args[0]}`
    }
  },

  formatters: {
    /**
     * Map %o to `util.inspect()`, all on a single line.
     */
    o (value) {
      createDebug.inspectOpts.colors = createDebug.useColors
      return util.inspect(value, createDebug.inspectOpts)
        .replace(/\s*\n\s*/g, ' ')
    },
    /**
     * Map %O to `util.inspect()`, allowing multiple lines if needed.
     */
    O (value) {
      createDebug.inspectOpts.colors = createDebug.useColors
      return util.inspect(value, createDebug.inspectOpts)
    }
  },

  /**
   * Build up the default `inspectOpts` object from the environment variables.
   *
   *   $ DEBUG_COLORS=no DEBUG_DEPTH=10 DEBUG_SHOW_HIDDEN=enabled node script.js
   */
  inspectOpts: Object.keys(process.env)
    .filter(key => /^debug_/i.test(key))
    .reduce((obj, key) => {
      // camel-case
      const prop = key
        .substring(6)
        .toLowerCase()
        .replace(/_([a-z])/g, (_, k) => k.toUpperCase())

      // coerce string value into JS value
      let value = process.env[key]
      if (/^(yes|on|true|enabled)$/i.test(value)) value = true
      else if (/^(no|off|false|disabled)$/i.test(value)) value = false
      else if (value === 'null') value = null
      else value = Number(value)

      obj[prop] = value
      return obj
    }, {}),

  /**
   * Load `namespaces`.
   *
   * @return {String} returns the previously persisted debug modes
   * @api private
   */
  load () {
    return process.env.DEBUG
  },

  /**
   * Invokes `util.format()` with the specified arguments and writes to stderr.
   */

  log () {
    return process.stderr.write(util.format.apply(util, arguments) + '\n')
  },

  /**
   * Save `namespaces`.
   *
   * @param {String} namespaces
   * @api private
   */

  save (namespaces) {
    if (namespaces == null) {
      // If you set a process.env field to null or undefined, it gets cast to the
      // string 'null' or 'undefined'. Just delete instead.
      delete process.env.DEBUG
    } else {
      process.env.DEBUG = namespaces
    }
  },

  /**
   * Is stdout a TTY? Colored output is enabled when `true`.
   */
  useColors () {
    return 'colors' in this.inspectOpts
      ? Boolean(this.inspectOpts.colors)
      : tty.isatty(process.stderr.fd)
  }
})

const initFunc = () => init(inject, 'node')
initFunc.inject = inject

export default initFunc
